File: ingest/embedder.py
"""② 청킹 + 임베딩 (SPEC §7.1).

청킹: 문단 단위, 목표 ~500자. 첫 청크에는 `title + summary` 를 앞에 붙여 문맥을 준다.
청크별로 `program_embeddings` 한 행이 되고, 질의 시 프로그램 단위 MAX(sim)로 집계한다
(벡터 자체를 pooling 하지 않는다 — SPEC §7.1).

Mock 임베딩은 **교차팀 확정 계약**이다. 웹팀이 TypeScript로 같은 알고리즘을 구현하고,
질의 벡터↔문서 벡터가 같은 공간에 있어야 데모가 동작한다. 아래 구현을 바꾸지 말 것::

    dim = 1024
    text → Unicode NFC 정규화 → lowercase → 코드포인트 시퀀스
    v = zeros(1024)
    인접한 코드포인트 쌍 (c1, c2) 마다:  idx = (c1*31 + c2) % 1024 ;  v[idx] += 1
    코드포인트가 2개 미만이면 v[0] = 1
    마지막에 L2 정규화

코사인 유사도가 문자 bigram 겹침을 재게 된다 — 데모 수준으로는 충분하다.
실 provider(voyage/openai)는 **W1에 한국어 성능 실측 후 확정** (SPEC §4, §12 리스크 2).
"""

from __future__ import annotations

import logging
import math
import re
import unicodedata
from dataclasses import dataclass
from typing import Sequence

from .config import EMBEDDING_DIM, Settings

try:  # pragma: no cover - 실 provider 호출 경로에서만 필요
    import httpx  # type: ignore
except ImportError:  # pragma: no cover
    httpx = None  # type: ignore[assignment]

log = logging.getLogger(__name__)

TARGET_CHUNK_CHARS = 500
MAX_CHUNK_CHARS = 900
MAX_CHUNKS = 20


# ---------------------------------------------------------------------- 청킹


def _split_long_paragraph(paragraph: str, limit: int) -> list[str]:
    """한 문단이 limit 를 넘으면 문장 경계로 자른다."""
    sentences = re.split(r"(?<=[.!?。])\s+|(?<=다\.)\s*", paragraph)
    sentences = [s.strip() for s in sentences if s and s.strip()]
    if not sentences:
        return [paragraph[:limit]]
    out: list[str] = []
    buf = ""
    for sentence in sentences:
        while len(sentence) > limit:  # 문장 하나가 이미 너무 길면 강제 절단
            out.append(sentence[:limit])
            sentence = sentence[limit:]
        if buf and len(buf) + len(sentence) + 1 > limit:
            out.append(buf)
            buf = sentence
        else:
            buf = f"{buf} {sentence}".strip()
    if buf:
        out.append(buf)
    return out


def chunk_text(
    body: str,
    *,
    title: str = "",
    summary: str = "",
    target: int = TARGET_CHUNK_CHARS,
) -> list[str]:
    """문단 기반 청킹. 첫 청크 앞에 title + summary 를 붙인다 (SPEC §7.1)."""
    body = (body or "").strip()
    paragraphs: list[str] = []
    for raw in re.split(r"\n\s*\n", body):
        raw = raw.strip()
        if not raw:
            continue
        if len(raw) > MAX_CHUNK_CHARS:
            paragraphs.extend(_split_long_paragraph(raw, target))
        else:
            paragraphs.append(raw)

    chunks: list[str] = []
    buf = ""
    for paragraph in paragraphs:
        if buf and len(buf) + len(paragraph) + 1 > target:
            chunks.append(buf)
            buf = paragraph
        else:
            buf = f"{buf}\n{paragraph}".strip()
    if buf:
        chunks.append(buf)

    header = "\n".join(p for p in (title.strip(), summary.strip()) if p)
    if not chunks:
        return [header] if header else []
    if header:
        chunks[0] = f"{header}\n{chunks[0]}"
    return chunks[:MAX_CHUNKS]


# ----------------------------------------------------------------- mock 임베딩


def mock_embed(text: str, dim: int = EMBEDDING_DIM) -> list[float]:
    """문자 bigram 해시 임베딩 — 웹팀 TypeScript 구현과 **바이트 단위로 동일해야 한다.**"""
    normalized = unicodedata.normalize("NFC", text or "").lower()
    codepoints = [ord(ch) for ch in normalized]
    vector = [0.0] * dim
    if len(codepoints) < 2:
        vector[0] = 1.0
    else:
        for first, second in zip(codepoints, codepoints[1:]):
            vector[(first * 31 + second) % dim] += 1.0
    norm = math.sqrt(sum(v * v for v in vector))
    if norm == 0.0:  # 이론상 도달 불가 — 방어적으로 결정론을 유지한다
        vector[0] = 1.0
        return vector
    return [v / norm for v in vector]


def cosine(a: Sequence[float], b: Sequence[float]) -> float:
    return sum(x * y for x, y in zip(a, b))


# ------------------------------------------------------------------- Embedder


@dataclass
class EmbeddingResult:
    chunks: list[str]
    vectors: list[list[float]]
    provider: str


class Embedder:
    """EMBEDDING_PROVIDER 로 구현을 고른다. 기본은 mock."""

    def __init__(self, settings: Settings | None = None, *, dim: int = EMBEDDING_DIM) -> None:
        self.settings = settings or Settings.from_env()
        self.dim = dim
        self.provider = "mock" if self.settings.mock_embeddings else self.settings.embedding_provider
        self.api_calls = 0

    # -- 실 provider 스텁 ------------------------------------------------
    # 엔드포인트·모델명·차원은 **W1에 한국어 성능 실측 후 확정**한다 (SPEC §4, §12).
    # 지금 확정처럼 적어두면 W1에서 조용히 어긋난다. 실측 전까지는 mock 이 기본이다.

    def _embed_voyage(self, texts: list[str]) -> list[list[float]]:  # pragma: no cover
        if httpx is None:
            raise RuntimeError("httpx 미설치 — voyage provider 사용 불가")
        response = httpx.post(
            "https://api.voyageai.com/v1/embeddings",
            headers={"Authorization": f"Bearer {self.settings.embedding_api_key}"},
            json={"model": "voyage-3", "input": texts, "input_type": "document"},
            timeout=60.0,
        )
        response.raise_for_status()
        return [row["embedding"] for row in response.json()["data"]]

    def _embed_openai(self, texts: list[str]) -> list[list[float]]:  # pragma: no cover
        if httpx is None:
            raise RuntimeError("httpx 미설치 — openai provider 사용 불가")
        response = httpx.post(
            "https://api.openai.com/v1/embeddings",
            headers={"Authorization": f"Bearer {self.settings.embedding_api_key}"},
            json={"model": "text-embedding-3-small", "input": texts, "dimensions": self.dim},
            timeout=60.0,
        )
        response.raise_for_status()
        return [row["embedding"] for row in response.json()["data"]]

    # -- 공개 API --------------------------------------------------------

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        if self.provider == "mock":
            return [mock_embed(t, self.dim) for t in texts]
        try:
            self.api_calls += 1
            if self.provider == "voyage":
                vectors = self._embed_voyage(texts)
            elif self.provider == "openai":
                vectors = self._embed_openai(texts)
            else:
                raise RuntimeError(f"알 수 없는 EMBEDDING_PROVIDER: {self.provider}")
        except Exception as exc:  # 임베딩 실패 시 mock 으로 강등 (배치 중단 금지)
            log.warning("임베딩 provider '%s' 실패 → mock 강등: %s", self.provider, exc)
            return [mock_embed(t, self.dim) for t in texts]
        for vector in vectors:
            if len(vector) != self.dim:
                raise ValueError(f"임베딩 차원 불일치: {len(vector)} != {self.dim}")
        return vectors

    def embed_program(self, *, title: str, summary: str, body: str) -> EmbeddingResult:
        chunks = chunk_text(body, title=title, summary=summary)
        return EmbeddingResult(chunks=chunks, vectors=self.embed_texts(chunks), provider=self.provider)

File: web/ingest/embedder.ts
import {
  EMBEDDING_DIM,
  mockEmbed,
  OPENAI_MODEL,
  VOYAGE_MODEL,
  type EmbeddingProvider,
} from "../src/lib/embedding";

export { OPENAI_MODEL };

export const TARGET_CHUNK_CHARS = 500;
export const MAX_CHUNK_CHARS = 900;
export const MAX_CHUNKS = 20;

function splitLongParagraph(paragraph: string, limit: number): string[] {
  const sentences = paragraph
    .split(/(?<=[.!?。])\s+|(?<=다\.)\s*/u)
    .map((sentence) => sentence.trim())
    .filter(Boolean);
  if (!sentences.length) return [paragraph.slice(0, limit)];

  const out: string[] = [];
  let buffer = "";
  for (let sentence of sentences) {
    while (sentence.length > limit) {
      out.push(sentence.slice(0, limit));
      sentence = sentence.slice(limit);
    }
    if (buffer && buffer.length + sentence.length + 1 > limit) {
      out.push(buffer);
      buffer = sentence;
    } else {
      buffer = `${buffer} ${sentence}`.trim();
    }
  }
  if (buffer) out.push(buffer);
  return out;
}

export function chunkText(
  body: string,
  options: { title?: string; summary?: string; target?: number } = {},
): string[] {
  const { title = "", summary = "", target = TARGET_CHUNK_CHARS } = options;
  const paragraphs: string[] = [];
  for (const raw of (body ?? "").trim().split(/\n\s*\n/u)) {
    const paragraph = raw.trim();
    if (!paragraph) continue;
    paragraphs.push(
      ...(paragraph.length > MAX_CHUNK_CHARS
        ? splitLongParagraph(paragraph, target)
        : [paragraph]),
    );
  }

  const chunks: string[] = [];
  let buffer = "";
  for (const paragraph of paragraphs) {
    if (buffer && buffer.length + paragraph.length + 1 > target) {
      chunks.push(buffer);
      buffer = paragraph;
    } else {
      buffer = `${buffer}\n${paragraph}`.trim();
    }
  }
  if (buffer) chunks.push(buffer);

  const header = [title.trim(), summary.trim()].filter(Boolean).join("\n");
  if (!chunks.length) return header ? [header] : [];
  if (header) chunks[0] = `${header}\n${chunks[0]}`;
  return chunks.slice(0, MAX_CHUNKS);
}

export type EmbeddingResult = {
  chunks: string[];
  vectors: number[][];
  provider: EmbeddingProvider;
};

export type EmbeddingSettings = {
  embeddingProvider?: string;
  embeddingApiKey?: string;
  mockEmbeddings?: boolean;
  embedding_provider?: string;
  embedding_api_key?: string;
  mock_embeddings?: boolean;
};

function settingsFromEnv(): {
  embeddingProvider: string;
  embeddingApiKey: string;
  mockEmbeddings: boolean;
} {
  return {
    embeddingProvider: (process.env.EMBEDDING_PROVIDER ?? "mock").toLowerCase(),
    embeddingApiKey: process.env.EMBEDDING_API_KEY ?? "",
    mockEmbeddings: process.env.MOCK_EMBEDDINGS === "1",
  };
}

export class Embedder {
  readonly provider: EmbeddingProvider;
  /**
   * 저장·비교에 쓰는 벡터 공간 식별자. provider 만으로는 같은 provider 안에서의
   * 모델 교체(voyage-3 → voyage-4-large)를 구분하지 못해, 옛 벡터가 새 질의 벡터와
   * 비교되며 유사도가 조용히 망가진다. 모델명을 포함해 재색인이 트리거되게 한다.
   */
  readonly vectorSpace: string;
  readonly apiKey: string;
  apiCalls = 0;

  constructor(settings: EmbeddingSettings = {}) {
    const env = settingsFromEnv();
    const requested = (
      settings.embeddingProvider ?? settings.embedding_provider ?? env.embeddingProvider
    ).toLowerCase();
    // Shared Settings also marks a missing real-provider key as mock_embeddings=true.
    // Do not honor that legacy downgrade unless the selected provider is actually mock.
    const forceMock = (settings.mockEmbeddings ?? env.mockEmbeddings) || requested === "mock";
    if (!forceMock && requested !== "mock" && requested !== "voyage" && requested !== "openai") {
      throw new Error(`unknown EMBEDDING_PROVIDER: ${requested}`);
    }
    this.provider = forceMock ? "mock" : (requested as EmbeddingProvider);
    this.vectorSpace =
      this.provider === "mock"
        ? "mock"
        : `${this.provider}:${this.provider === "voyage" ? VOYAGE_MODEL : OPENAI_MODEL}`;
    this.apiKey = settings.embeddingApiKey ?? settings.embedding_api_key ?? env.embeddingApiKey;
  }

  private async embedRemote(texts: string[]): Promise<number[][]> {
    if (!this.apiKey) throw new Error(`EMBEDDING_API_KEY is required for ${this.provider}`);
    this.apiCalls++;
    const response = await fetch(
      this.provider === "voyage"
        ? "https://api.voyageai.com/v1/embeddings"
        : "https://api.openai.com/v1/embeddings",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(
          this.provider === "voyage"
            ? {
                model: VOYAGE_MODEL,
                input: texts,
                input_type: "document",
                output_dimension: EMBEDDING_DIM,
              }
            : {
                model: OPENAI_MODEL,
                input: texts,
                dimensions: EMBEDDING_DIM,
              },
        ),
        signal: AbortSignal.timeout(60_000),
      },
    );
    if (!response.ok) throw new Error(`embedding API ${response.status}`);
    const data = (await response.json()) as { data?: Array<{ embedding?: unknown }> };
    if (!Array.isArray(data.data) || data.data.length !== texts.length) {
      throw new Error("invalid embedding response count");
    }
    return data.data.map(({ embedding }) => {
      if (
        !Array.isArray(embedding) ||
        embedding.length !== EMBEDDING_DIM ||
        !embedding.every((value) => typeof value === "number" && Number.isFinite(value)) ||
        !embedding.some((value) => value !== 0)
      ) {
        throw new Error("invalid embedding response vector");
      }
      return embedding as number[];
    });
  }

  async embedTexts(texts: string[]): Promise<number[][]> {
    if (!texts.length) return [];
    if (this.provider === "mock") return texts.map((text) => Array.from(mockEmbed(text)));
    // Never write mock vectors under a real provider: let the record transaction roll back.
    return this.embedRemote(texts);
  }

  async embedProgram({
    title,
    summary,
    body,
  }: {
    title: string;
    summary: string;
    body: string;
  }): Promise<EmbeddingResult> {
    const chunks = chunkText(body, { title, summary });
    return { chunks, vectors: await this.embedTexts(chunks), provider: this.provider };
  }
}
